include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+			ix = choseRight ? right:(ix+1u);
+			weight *= choseRight ? (1.0-w):w;
+		}
+		else 
+#endif //OP_BLEND
+		{
+#ifdef OP_SET_GEOM_NORMAL
+			if (op==OP_SET_GEOM_NORMAL)
+			{
+				instr_execute_SET_GEOM_NORMAL();
+			} else 
+#endif //OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			if (op==OP_BUMPMAP)
+			{
+				mat2x4 srcs = instr_fetchSrcRegs(instr);
+				instr_execute_BUMPMAP(instr, srcs);
+			} else
+#endif //OP_BUMPMAP
+			{}
+			ix = ix + 1u;
+		}
+
+		instr = irr_glsl_MC_fetchInstr(stream.offset+ix);
+		op = instr_getOpcode(instr);
+	}
+
+	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+	params_t params = instr_getParameters(instr, bsdf_data);
+	//speculatively
+	const float ax = params_getAlpha(params);
+	const float ax2 = ax*ax;
+	const float ay = params_getAlphaV(params);
+	const float ay2 = ay*ay;
+	const mat2x3 ior = bsdf_data_decodeIoR(bsdf_data,op);
+	const bool is_bsdf = !op_isBRDF(op);
+	const vec3 refl = params_getReflectance(params);
+
+#ifndef NO_TWOSIDED
+	bool ts_flag = false;
+	handleTwosided_interactionOnly(ts_flag, instr);
+#endif //NO_TWOSIDED
+
+	const bool is_plastic = op==OP_PLASTIC;
+
+#ifdef OP_DIFFUSE
+
+#define OP_DIFFUSE_ALIAS OP_DIFFUSE
+
+#else
+
+#define OP_DIFFUSE_ALIAS 0xdeadbeefu
+
+#endif
+
+	float pdf;
+	vec3 rem;
+	uint ndf = instr_getNDF(instr);
+	irr_glsl_LightSample s;
+	irr_glsl_AnisotropicMicrofacetCache uf;
+
+	// the if/else statements below cannot be done as switch/case because it checks both `op` and `ndf`
+#ifdef OP_PLASTIC
+	float plastic_spec_weight;
+	if (is_plastic) {
+		vec3 fresnel = irr_glsl_fresnel_dielectric(ior[0],currInteraction.isotropic.NdotV);
+		float ws = dot(fresnel, CIE_XYZ_Luma_Y_coeffs);
+		bool choseDiffuse = u.z>=ws;
+		op = choseDiffuse ? OP_DIFFUSE_ALIAS : op;
+		plastic_spec_weight = ws;
+	}
+#endif //OP_PLASTIC
+
+#if defined(OP_DIFFUSE) || defined(OP_PLASTIC)
+	if (op==OP_DIFFUSE_ALIAS) {
+		s = irr_glsl_oren_nayar_cos_generate(currInteraction, u.xy, ax2);
+		uf = getSmoothMicrofacetCache(currInteraction.isotropic.NdotV, s.NdotL);
+		rem = refl*irr_glsl_oren_nayar_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, ax2);
+	} else 
+#endif //OP_DIFFUSE
+
+#ifdef OP_DIFFTRANS
+	if (op == OP_DIFFTRANS) {
+		s = irr_glsl_lambertian_transmitter_cos_generate(currInteraction, u);
+		uf = getSmoothMicrofacetCache(currInteraction.isotropic.NdotV, s.NdotL);//TODO?
+		rem = refl*irr_glsl_lambertian_transmitter_cos_remainder_and_pdf(pdf, s);
+	} else
+#endif //OP_DIFFTRANS
+
+#ifdef OP_THINDIELECTRIC
+	if (op == OP_THINDIELECTRIC) {
+		const vec3 luminosityContributionHint = CIE_XYZ_Luma_Y_coeffs;
+		s = irr_glsl_thin_smooth_dielectric_cos_generate(currInteraction, u, ior[0]*ior[0], luminosityContributionHint);
+		uf = getSmoothMicrofacetCache(currInteraction.isotropic.NdotV, s.NdotL);
+		rem = irr_glsl_thin_smooth_dielectric_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, ior[0]*ior[0], luminosityContributionHint);
+		pdf = 1.0;
```

Comment:
why did you set PDF to 1.0 ? It should be infinite!

Reply:
PDF of a delta BSDF as a generator is infinite, as a evaluator its 0
(the probability density around the only value you can generate is infinite, but if the exact generator isnt used then its impossible to obtain the exact value).